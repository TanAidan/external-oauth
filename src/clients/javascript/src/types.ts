declare namespace Types {
  export interface ConnectOptions {
    authId?: string
    setupId?: string
    params?: any
  }

  export interface ConnectSuccess {
    authId: string
  }

  export interface ConnectError extends Error {}

  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
