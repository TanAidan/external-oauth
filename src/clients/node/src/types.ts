declare namespace Types {
  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
