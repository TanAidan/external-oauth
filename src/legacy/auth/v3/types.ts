import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
