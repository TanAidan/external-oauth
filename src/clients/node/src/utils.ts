import Types from './types'

/**
 * Helper to build a new URL from different params
 */
export function toURL(origin: string, baseURL: string, endpoint: string, queryString?: Types.RequestQueryString): URL {
  const removeLeadingSlash = (text: string) => {
    return text.replace(/^\//, '')
  }

  const removeTrailingSlash = (text: string) => {
    return text.replace(/\/$/, '')
  }

  const urlParts: string[] = []
  urlParts.push(removeTrailingSlash(origin))
  urlParts.push(removeLeadingSlash(removeTrailingSlash(baseURL)))
  urlParts.push(removeLeadingSlash(endpoint))

  const url = new URL(urlParts.join('/'))

  if (queryString) {
    Object.keys(queryString).forEach(key => url.searchParams.append(key, String(queryString[key])))
  }

  return url
}

/**
 * Helper to remove all undefined keys
 * @param obj {object}
 */
export function cleanHeaders(obj: Record<string, any>) {
  return Object.keys(obj).reduce((acc, key: string) => {
    if (obj[key] !== undefined) {
      acc[key] = obj[key]
    }
    return acc
  }, {} as Record<string, any>)
}
