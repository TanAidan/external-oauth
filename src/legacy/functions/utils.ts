// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
