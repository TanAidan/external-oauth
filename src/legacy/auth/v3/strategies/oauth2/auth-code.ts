import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getCodeRedirectURL, getTokenWithCode } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'
import { AuthenticationFailed } from '../../errors'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  const callbackURL = process.env.AUTH_CALLBACK_URL || `${req.protocol}://${req.get('host')}/auth/callback`
  const {
    credentials: { clientId, clientSecret },
    scopes = []
  } = req.setupDetails
  const { code, error } = req.query
  const {
    authorizationURL,
    authorizationMethod,
    authorizationParams,
    bodyFormat,
    config,
    tokenParams,
    tokenURL
  } = req.integrationConfig

  if (error) {
    throw AuthenticationFailed.fromOAuthRequest(req, undefined)
  }

  if (code) {
    const tokenResult = await getTokenWithCode({
      authorizationMethod,
      bodyFormat,
      clientId,
      clientSecret,
      code,
      tokenParams,
      tokenURL,
      callbackURL
    })

    // console.log('tokenResult', tokenResult)
    req.credentials = responseToCredentials(tokenResult)
    // console.log('credentials', req.credentials)
    // console.log('decodedToken', tokenResult.decodedResponse)
    req.tokenResponse = tokenResult.decodedResponse

    return next()
  }

  // const { scope = [], state = 'none' } = config || {}
  const { state = 'none' } = config || {}

  const redirectURL = getCodeRedirectURL({
    authorizationParams,
    authorizationURL,
    clientId,
    state,
    scope: scopes || config?.scope || [],
    callbackURL
  })

  res.redirect(redirectURL)
})
