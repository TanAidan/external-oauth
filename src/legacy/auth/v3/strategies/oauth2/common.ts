import jwt from 'jsonwebtoken'

import { TokenResult } from '../../../clients/oauth2'
import { IOAuth2Credentials } from '../../types'

export const NO_VALUE = 'non'

export const responseToCredentials = ({ accessToken, refreshToken, idToken, expiresIn }: TokenResult) => {
  const credentials: IOAuth2Credentials = {
    accessToken,
    refreshToken: refreshToken || NO_VALUE,
    idToken: idToken || NO_VALUE,
    expiresIn: expiresIn || 0
  }

  if (idToken) {
    credentials.idTokenJwt = getIdTokenJwt(idToken)
  }

  return credentials
}

export const getIdTokenJwt = (idToken?: string) => {
  if (!idToken) {
    return undefined
  }

  try {
    return jwt.decode(idToken) || undefined
  } catch (e) {
    return undefined
  }
}
