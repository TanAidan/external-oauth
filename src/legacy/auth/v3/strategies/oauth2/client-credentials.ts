import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getTokenWithClientCredentials } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, _res: Response, next: NextFunction) => {
  const { authorizationMethod, bodyFormat, config, tokenURL } = req.integrationConfig
  const { scope = [] } = config || {}
  const { clientId, clientSecret } = req.setupDetails.credentials

  const tokenResult = await getTokenWithClientCredentials({
    authorizationMethod,
    bodyFormat,
    clientId,
    clientSecret,
    scope,
    tokenURL
  })

  req.credentials = responseToCredentials(tokenResult)
  req.tokenResponse = tokenResult.decodedResponse

  next()
})
