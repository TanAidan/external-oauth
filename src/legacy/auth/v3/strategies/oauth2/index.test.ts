import timekeeper from 'timekeeper'
import { RequestHandler } from 'express'

import { MiddlewareTestHarness } from '../../../../../../tests/utils'
import { authenticate, fetchAuthDetails, CLOCK_DRIFT_MS, GrantType } from '.'
import { TAuthenticateRequest, EAuthType } from '../../types'
// import { getSetupDetails, getAuth, updateAuthV3 } from '../../../../clients/integrations'
// import { mocked } from 'ts-jest/utils'
import Integration from '../../../../functions/integration'
import { authenticate as authCodeAuthenticate } from './auth-code'
import { authenticate as clientCredentialsAuthenticate } from './client-credentials'
import {
  getTokenWithRefreshToken,
  getTokenWithClientCredentials,
  AuthorizationMethod,
  BodyFormat
} from '../../../clients/oauth2'
import { responseToCredentials, getIdTokenJwt } from './common'
// import { expandAuthConfig } from '../../../../api-config/auth-config'

jest.mock('../../../../api-config/auth-config')
jest.mock('../../../../clients/integrations')
jest.mock('../../../../clients/oauth2')
jest.mock('./auth-code')
jest.mock('./client-credentials')
jest.mock('./common')

const requiredIntegrationConfig = {
  authType: EAuthType.OAuth2,
  authorizationURL: 'https://example.com/auth',
  tokenURL: 'https://${authDomain}.example.com/token'
}

describe('authenticate', () => {
  const testResponse = { test: 'response' }

  const mockAuth = (test: MiddlewareTestHarness<TAuthenticateRequest>, middleware: RequestHandler) => {
    // mocked(middleware).mockImplementationOnce((req: TAuthenticateRequest, res: Response, next: NextFunction) => {
    //   expect(req).toBe(test.req)
    //   res.json(testResponse).end()
    //   next()
    // })
  }

  const setup = (grantType: GrantType | undefined) =>
    new MiddlewareTestHarness<TAuthenticateRequest>({
      configureRequest: req => {
        req.integrationConfig = {
          ...requiredIntegrationConfig,
          grantType
        }
      },
      testMiddleware: authenticate
    })

  describe('when the grantType is not specified', () => {
    it('passes through to the auth code authenticate middleware', async () => {
      const test = setup(undefined)
      mockAuth(test, authCodeAuthenticate)

      await test
        .get()
        .expect(testResponse)
        .expect(200)
    })
  })

  describe('when the grantType is AuthCode', () => {
    it('passes through to the auth code authenticate middleware', async () => {
      const test = setup(GrantType.AuthCode)
      mockAuth(test, authCodeAuthenticate)

      await test
        .get()
        .expect(testResponse)
        .expect(200)
    })
  })

  describe('when the grantType is ClientCredentials', () => {
    it('passes through to the client credentials authenticate middleware', async () => {
      const test = setup(GrantType.ClientCredentials)
      mockAuth(test, clientCredentialsAuthenticate)

      await test
        .get()
        .expect(testResponse)
        .expect(200)
    })
  })

  describe('when the grantType is any other value', () => {
    it('throws an InvalidGrantType error', async () => {
      const test = setup('invalid' as GrantType)

      await test.get().expect(422)

      expect(test.err).toMatchSnapshot()
    })
  })
})

describe('fetchAuthDetails', () => {
  const now = new Date(Date.UTC(2019, 0, 1, 0, 0, 30) - CLOCK_DRIFT_MS)
  // const updatedAt = new Date(Date.UTC(2019, 0, 1, 0, 0, 0)).valueOf()

  const expiresInNever = 0
  const expiresInNotExpired = 30
  const expiresInExpired = 29

  const oAuthClientId = 'test-oauth-cient-id'
  const clientSecret = 'test-client-secret'
  const accessToken = 'test-access-token'
  // const refreshToken = 'test-refresh-token'
  // const idToken = 'test-id-token'
  const setupId = 'test-setup-id'

  // const connectParams = { connect: 'connect-param' }
  const callbackParams = { param: 'value' }
  const tokenResponse = { body: { hey: 'there' } }

  const newAccessToken = 'new-access-token'
  // const newRefreshToken = 'new-refresh-token'
  // const newIdToken = 'new-id-token'
  // const newExpiresIn = 42

  const log = jest.fn()

  const paramsBase: any = {
    buid: 'test-alias',
    scopedUserDataTableName: 'test-scoped-data-table',
    servicesTableName: 'test-services-table',
    environmentIdentifier: 'test-env-id',
    integration: new Integration('test-buid'),
    authId: 'test-authId',
    connectParams: {
      authDomain: 'auth22'
    },
    logger: { log }
  }

  const params = {
    ...paramsBase,
    setupId
  }

  const setup = ({}: // expiresIn,
  // withCallbackParams = false,
  // withCredentials = true,
  // withClientCredentials = true,
  // withRefreshToken = true,
  // withTokenResponse = false,
  // idTokenValue = idToken,
  // idTokenJwt
  {
    expiresIn: number
    withCallbackParams?: boolean
    withCredentials?: boolean
    withClientCredentials?: boolean
    withRefreshToken?: boolean
    withTokenResponse?: boolean
    idTokenValue?: string
    idTokenJwt?: any
  }) => {
    // const effectiveRefreshToken = withRefreshToken ? refreshToken : 'non'
    // const credentials = {
    //   setupId,
    //   connectParams,
    //   expiresIn,
    //   updatedAt,
    //   accessToken,
    //   idTokenJwt,
    //   callbackParamsJSON: withCallbackParams ? JSON.stringify(callbackParams) : undefined,
    //   clientId: withClientCredentials ? oAuthClientId : undefined,
    //   clientSecret: withClientCredentials ? clientSecret : undefined,
    //   idToken: idTokenValue,
    //   refreshToken: effectiveRefreshToken,
    //   tokenResponseJSON: withTokenResponse ? JSON.stringify(tokenResponse) : undefined
    // }

    // mocked(getSetupDetails)
    //   .mockResolvedValueOnce({ clientSecret, clientId: oAuthClientId })
    //   .mockClear()
    // mocked(getAuth)
    //   .mockResolvedValueOnce(withCredentials ? credentials : undefined)
    //   .mockClear()
    // mocked(updateAuthV3).mockClear()
    log.mockClear()
    // mocked(getIdTokenJwt).mockClear()
  }

  beforeAll(() => {
    timekeeper.freeze(now)
  })

  afterAll(() => {
    timekeeper.reset()
  })

  describe('when the token never expires', () => {
    it('returns the details from the auth credentials', async () => {
      setup({ expiresIn: expiresInNever })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth).toMatchSnapshot()
      // expect(getSetupDetails).not.toHaveBeenCalled()
      // expect(getAuth).toMatchSnapshot()
    })

    it('decodes any token call response JSON data', async () => {
      setup({ expiresIn: expiresInNever, withTokenResponse: true })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth.tokenResponse).toEqual(tokenResponse)
    })

    it('looks up the clientId and clientSecret if not stored with the credentials', async () => {
      setup({ expiresIn: expiresInNever, withClientCredentials: false })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      // expect(getSetupDetails).toHaveBeenCalled()
      expect(auth).toMatchObject({ clientSecret, clientId: oAuthClientId })
    })
  })

  describe('when the token has NOT expired', () => {
    const storedIdTokenJwt = { realmid: '123456789' }
    const decodedIdTokenJwt = { realmid: '999666222' }

    it('returns the details from the auth credentials', async () => {
      setup({ expiresIn: expiresInNotExpired })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth).toMatchSnapshot()
      // expect(getSetupDetails).not.toHaveBeenCalled()
      // expect(getAuth).toMatchSnapshot()
    })

    describe('when there is a JWT Id Token stored', () => {
      it('returns it', async () => {
        setup({ expiresIn: expiresInNotExpired, idTokenJwt: storedIdTokenJwt })

        const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

        expect(getIdTokenJwt).not.toHaveBeenCalled()
        expect(auth.idTokenJwt).toEqual(storedIdTokenJwt)
      })
    })

    describe('when there is NOT a JWT Id Token stored', () => {
      it('attempts to decode the Id Token and returns the result', async () => {
        // mocked(getIdTokenJwt).mockReturnValueOnce(decodedIdTokenJwt)
        setup({ expiresIn: expiresInNotExpired })

        const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

        expect(auth.idTokenJwt).toEqual(decodedIdTokenJwt)
      })
    })

    it('decodes any token call response JSON data', async () => {
      setup({ expiresIn: expiresInNotExpired, withTokenResponse: true })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth.tokenResponse).toEqual(tokenResponse)
    })

    it('returns any callback params data', async () => {
      setup({ expiresIn: expiresInNotExpired, withCallbackParams: true })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth.callbackParams).toEqual(callbackParams)
    })
  })

  describe('when the token has expired', () => {
    // const expandedIntegrationConfig = {
    //   ...requiredIntegrationConfig,
    //   tokenURL: 'https://auth22.example.com/token'
    // }

    const refreshURL = 'https://example.com/refresh'

    const newTokenResponse = { body: { new: 'data' } }
    const refreshResponse = { accessToken: newAccessToken, decodedResponse: newTokenResponse }

    // const newCredentials = {
    //   accessToken: newAccessToken,
    //   idToken: newIdToken,
    //   refreshToken: newRefreshToken,
    //   expiresIn: newExpiresIn
    // }

    beforeAll(() => {
      // mocked(expandAuthConfig).mockReturnValue(expandedIntegrationConfig)
      // mocked(getTokenWithClientCredentials).mockReturnValue(refreshResponse)
      // mocked(getTokenWithRefreshToken).mockReturnValue(refreshResponse)
      // mocked(responseToCredentials).mockReturnValue(newCredentials)
    })

    beforeEach(() => {
      // mocked(expandAuthConfig).mockClear()
      // mocked(getTokenWithClientCredentials).mockClear()
      // mocked(getTokenWithRefreshToken).mockClear()
      // mocked(responseToCredentials).mockClear()
    })

    it('refreshes the token using the auth credentials and setup details', async () => {
      setup({ expiresIn: expiresInExpired })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      // expect(getSetupDetails).toMatchSnapshot()
      // expect(getAuth).toMatchSnapshot()
      expect(responseToCredentials).toHaveBeenCalledWith(refreshResponse)
      expect(getTokenWithRefreshToken).toMatchSnapshot()
      expect(auth).toMatchSnapshot()
    })

    it('updates the stored auth credentials with the refreshed values', async () => {
      setup({ expiresIn: expiresInExpired })

      await fetchAuthDetails(params, requiredIntegrationConfig)

      // expect(updateAuthV3).toMatchSnapshot()
    })

    it('keeps any callback params that were previously stored', async () => {
      setup({ expiresIn: expiresInExpired, withCallbackParams: true })

      await fetchAuthDetails(params, requiredIntegrationConfig)

      // expect(updateAuthV3).toMatchSnapshot()
    })

    it('uses the refreshURL configuration option when specified', async () => {
      const integrationConfig = {
        ...requiredIntegrationConfig,
        refreshURL
      }

      setup({ expiresIn: expiresInExpired })
      // mocked(expandAuthConfig).mockReturnValue(integrationConfig)

      await fetchAuthDetails(params, integrationConfig)

      expect(getTokenWithRefreshToken).toHaveBeenCalledWith(expect.objectContaining({ tokenURL: refreshURL }))
    })

    it('returns the token call response JSON data, merged with the previous value', async () => {
      setup({ expiresIn: expiresInExpired, withTokenResponse: true })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth.tokenResponse).toEqual({ body: { hey: 'there', new: 'data' } })
    })

    it('returns any callback params data', async () => {
      setup({ expiresIn: expiresInExpired, withCallbackParams: true })

      const auth = await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(auth.callbackParams).toEqual(callbackParams)
    })

    describe('when there is no refresh token', () => {
      describe('when the grant type is Client Credentials', () => {
        it('gets a new token using the client credentials', async () => {
          const integrationConfig = {
            ...requiredIntegrationConfig,
            authorizationMethod: AuthorizationMethod.Header,
            bodyFormat: BodyFormat.Form,
            grantType: GrantType.ClientCredentials,
            config: {
              scope: ['read', 'write']
            }
          }

          // const expandedIntegrationConfig = {
          //   ...integrationConfig,
          //   tokenURL: 'https://auth22.example.com/token'
          // }

          setup({ expiresIn: expiresInExpired, withRefreshToken: false })
          // mocked(expandAuthConfig).mockReturnValueOnce(expandedIntegrationConfig)

          const auth = await fetchAuthDetails(params, integrationConfig)

          expect(getTokenWithClientCredentials).toMatchSnapshot()
          expect(auth).toMatchSnapshot()
        })
      })

      describe('when the grant type is NOT Client Credentials', () => {
        it('raises an AccessTokenExpired error', async () => {
          setup({ expiresIn: expiresInExpired, withRefreshToken: false })

          await expect(fetchAuthDetails(params, requiredIntegrationConfig)).rejects.toMatchSnapshot()
        })
      })
    })
  })

  describe("when the setupId used when connecting doesn't match the setupId sent by the user in the request", () => {
    it('throws an InconsistentSetupId error', async () => {
      setup({ expiresIn: expiresInNever })

      const params = { ...paramsBase, setupId: 'other-setup-id', setupIdFromRequest: true }

      await expect(fetchAuthDetails(params, requiredIntegrationConfig)).rejects.toMatchSnapshot()
    })
  })

  describe("when the setupId used when connecting doesn't match the default setupId", () => {
    it('logs a warning message about inconsistent setupId', async () => {
      setup({ expiresIn: expiresInNever })

      const params = { ...paramsBase, setupId: 'other-setup-id', setupIdFromRequest: false }
      await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(log).toHaveBeenCalledWith(
        // tslint:disable-next-line:max-line-length
        'Warning: setupId inconsistency - setupId used for authorization setup differs from latest setupId stored in API dashboard'
      )
    })
  })

  describe('when no setupId parameter is present', () => {
    it("doesn't raise an error", async () => {
      setup({ expiresIn: expiresInNever })

      const auth = await fetchAuthDetails(paramsBase, requiredIntegrationConfig)

      expect(auth).toMatchObject({ accessToken })
    })
  })

  describe('when no auth details can be found for the authId', () => {
    it('raises an InvalidAuthId error', async () => {
      setup({ expiresIn: expiresInNever, withCredentials: false })

      await expect(fetchAuthDetails(paramsBase, requiredIntegrationConfig)).rejects.toMatchInlineSnapshot(`
[InvalidAuthId: No auth details were found using authId 'test-authId' for API 'test-alias'

Please try again with a valid authId or connect with this authId]
`)
    })
  })
})
