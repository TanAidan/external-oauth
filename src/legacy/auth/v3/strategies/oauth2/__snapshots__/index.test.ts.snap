// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate when the grantType is any other value throws an InvalidGrantType error 1`] = `[InvalidGrantType: Invalid grantType 'invalid'. Please check the API's configuration. Allowed values are 'authorization_code', 'client_credentials']`;

exports[`fetchAuthDetails when the setupId used when connecting doesn't match the setupId sent by the user in the request throws an InconsistentSetupId error 1`] = `[InconsistentSetupId: The provided setupId does not match the setupId used when connecting. Please reconnect using this setupId, or retry this call with the setupId corresponding to the authId]`;

exports[`fetchAuthDetails when the token has NOT expired returns the details from the auth credentials 1`] = `
Object {
  "accessToken": "test-access-token",
  "callbackParams": undefined,
  "clientId": "test-oauth-cient-id",
  "clientSecret": "test-client-secret",
  "connectParams": Object {
    "connect": "connect-param",
  },
  "expiresIn": 30,
  "idToken": "test-id-token",
  "idTokenJwt": undefined,
  "refreshToken": "test-refresh-token",
  "tokenResponse": undefined,
  "updatedAt": 1546300800000,
}
`;

exports[`fetchAuthDetails when the token has NOT expired returns the details from the auth credentials 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-authId",
        "buid": "test-buid",
        "clientId": "test-env-id",
        "servicesTableName": "test-services-table",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`fetchAuthDetails when the token has expired refreshes the token using the auth credentials and setup details 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "clientId": "ClientID",
        "clientSecret": "clientSecret",
        "idToken": "abc",
        "refreshToken": "abc",
        "tokenURL": "https://auth22.example.com/token",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "accessToken": "new-access-token",
        "decodedResponse": Object {
          "body": Object {
            "new": "data",
          },
        },
      },
    },
  ],
}
`;

exports[`fetchAuthDetails when the token has expired refreshes the token using the auth credentials and setup details 2`] = `
Object {
  "accessToken": "new-access-token",
  "callbackParams": Object {},
  "clientId": "clientId",
  "clientSecret": "ClientSecret",
  "connectParams": Object {},
  "expiresIn": 42,
  "idToken": "new-id-token",
  "idTokenJwt": undefined,
  "refreshToken": "new-refresh-token",
  "tokenResponse": Object {
    "body": Object {
      "new": "data",
    },
  },
  "updatedAt": 1546300770000,
}
`;

exports[`fetchAuthDetails when the token has expired when there is no refresh token when the grant type is Client Credentials gets a new token using the client credentials 1`] = `[MockFunction]`;

exports[`fetchAuthDetails when the token has expired when there is no refresh token when the grant type is Client Credentials gets a new token using the client credentials 2`] = `
Object {
  "accessToken": "new-access-token",
  "callbackParams": Object {},
  "clientId": "clientId",
  "clientSecret": "ClientSecret",
  "connectParams": Object {},
  "expiresIn": 42,
  "idToken": "new-id-token",
  "idTokenJwt": undefined,
  "refreshToken": "new-refresh-token",
  "tokenResponse": Object {
    "body": Object {
      "new": "data",
    },
  },
  "updatedAt": 1546300770000,
}
`;

exports[`fetchAuthDetails when the token has expired when there is no refresh token when the grant type is NOT Client Credentials raises an AccessTokenExpired error 1`] = `
[AccessTokenExpired: The access token associated with your Auth Id has expired and token refreshing is not supported for this API. Please reconnect to obtain a new access token.

Consult the API provider's documentation to determine whether refresh tokens can be enabled for this API]
`;

exports[`fetchAuthDetails when the token never expires returns the details from the auth credentials 1`] = `
Object {
  "accessToken": "test-access-token",
  "callbackParams": undefined,
  "clientId": "test-oauth-cient-id",
  "clientSecret": "test-client-secret",
  "connectParams": Object {
    "connect": "connect-param",
  },
  "expiresIn": 0,
  "idToken": "test-id-token",
  "idTokenJwt": undefined,
  "refreshToken": "test-refresh-token",
  "tokenResponse": undefined,
  "updatedAt": 1546300800000,
}
`;

exports[`fetchAuthDetails when the token never expires returns the details from the auth credentials 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-authId",
        "buid": "test-buid",
        "clientId": "test-env-id",
        "servicesTableName": "test-services-table",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
